File: android/app/src/main/kotlin/com/example/drlogger/network/ApiService.kt
package com.example.drlogger.network

import com.example.drlogger.models.*
import okhttp3.MultipartBody
import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.Multipart
import retrofit2.http.POST
import retrofit2.http.PUT
import retrofit2.http.Part
import retrofit2.http.Query
import retrofit2.http.Url
import retrofit2.Response


interface ApiService {

    @GET("ping")
    suspend fun pingServer(): String

    @POST("v1/upload-session")
    suspend fun uploadSession(@Body request: SessionUploadRequest): ApiResponse

    @POST("v1/get-presigned-url")
    suspend fun getPresignedUrl(@Body request: ChunkUploadRequest): PresignedUrlResponse

    @PUT
    suspend fun uploadChunkToUrl(
        @Url presignedUrl: String,
        @Body fileBytes: ByteArray
    ): Void

    @Multipart
    @POST("https://file.io/")
    suspend fun uploadFile(
        @Part file: MultipartBody.Part,
        @Query("expires") expires: String = "1w"
    ): Response<FileIoResponse>


    @POST("v1/notify-chunk-uploaded")
    suspend fun notifyChunkUploaded(@Body request: ChunkUploadNotificationRequest): ApiResponse
}

File: ios/Runner/Swift/Network/ApiService.swift
import Foundation

class ApiService {
    static let shared = ApiService()
    
    private init() {}
    
    // MARK: - Ping Server (GET, no body)
    func pingServer(completion: @escaping (Result<String, Error>) -> Void) {
        ApiClient.shared.makeRequest(endpoint: "ping") { result in
            switch result {
            case .success(let data):
                completion(.success(String(data: data, encoding: .utf8) ?? ""))
            case .failure(let error):
                completion(.failure(error))
            }
        }
    }
    
    // MARK: - Upload Session
    func uploadSession(request: SessionUploadRequest, completion: @escaping (Result<ApiResponse, Error>) -> Void) {
        ApiClient.shared.makeRequest(endpoint: "v1/upload-session", method: "POST", body: request) { result in
            switch result {
            case .success(let data):
                do {
                    let response = try JSONDecoder().decode(ApiResponse.self, from: data)
                    completion(.success(response))
                } catch {
                    completion(.failure(error))
                }
            case .failure(let error):
                completion(.failure(error))
            }
        }
    }
    
    // MARK: - Get Presigned URL
    func getPresignedUrl(request: ChunkUploadRequest, completion: @escaping (Result<PresignedUrlResponse, Error>) -> Void) {
        ApiClient.shared.makeRequest(endpoint: "v1/get-presigned-url", method: "POST", body: request) { result in
            switch result {
            case .success(let data):
                do {
                    let response = try JSONDecoder().decode(PresignedUrlResponse.self, from: data)
                    completion(.success(response))
                } catch {
                    completion(.failure(error))
                }
            case .failure(let error):
                completion(.failure(error))
            }
        }
    }
    
    // MARK: - Upload Chunk to Presigned URL
    func uploadChunkToUrl(url: String, fileBytes: Data, completion: @escaping (Result<Void, Error>) -> Void) {
        guard let url = URL(string: url) else { return }
        var request = URLRequest(url: url)
        request.httpMethod = "PUT"
        request.httpBody = fileBytes
        request.setValue("application/octet-stream", forHTTPHeaderField: "Content-Type")
        
        let task = URLSession.shared.dataTask(with: request) { _, _, error in
            if let error = error {
                completion(.failure(error))
            } else {
                completion(.success(()))
            }
        }
        task.resume()
    }
    
    // MARK: - Notify Chunk Uploaded
    func notifyChunkUploaded(request: ChunkUploadNotificationRequest, completion: @escaping (Result<ApiResponse, Error>) -> Void) {
        ApiClient.shared.makeRequest(endpoint: "v1/notify-chunk-uploaded", method: "POST", body: request) { result in
            switch result {
            case .success(let data):
                do {
                    let response = try JSONDecoder().decode(ApiResponse.self, from: data)
                    completion(.success(response))
                } catch {
                    completion(.failure(error))
                }
            case .failure(let error):
                completion(.failure(error))
            }
        }
    }
}
